GenQuery2 docs and tweak to syslog docs (main) (#278)
irods/irods_docs: docs/system_overview/genquery.md

# iRODS General Query (GenQuery)

GenQuery is the SQL-like syntax used to query the iRODS catalog. It is used throughout the server to accomplish everything required of the system from resource hierarchy resolution to enforcing permissions, as well as directly by users via `iquest` and other clients.

> This page describes syntax for **GenQuery** because it has existed for almost as long as iRODS has existed and is used extensively throughout existing iRODS clients and servers. A more modern solution is now available in the iRODS server called [**GenQuery2**](#genquery2) which will eventually replace the legacy GenQuery implementation.

Reviewer: ```suggestion
This page describes syntax for **GenQuery** which is used extensively throughout iRODS clients and servers. A more modern solution is now available in the iRODS server called [**GenQuery2**](#genquery2) which will eventually replace the original GenQuery(1) implementation.
```

Author: Done.

Used GenQuery1 instead of GenQuery(1) since GenQuery1 is referenced throughout the doc.

| DATA_ACCESS_TYPE | DATA_ACCESS_PERM_ID |
| RESC_LOC | RESC_HOSTNAME |
| RULE_EXEC_ADDRESS | DELAY_RULE_EXE_ADDRESS |
| RULE_EXEC_ESTIMATED_EXE_TIME | DELAY_RULE_ESTIMATED_EXE_TIME |
| RULE_EXEC_FREQUENCY | DELAY_RULE_EXE_FREQUENCY |
| RULE_EXEC_ID | DELAY_RULE_ID |
| RULE_EXEC_LAST_EXE_TIME | DELAY_RULE_LAST_EXE_TIME |
| RULE_EXEC_NAME | DELAY_RULE_NAME |
| RULE_EXEC_NOTIFICATION_ADDR | DELAY_RULE_NOTIFICATION_ADDR |
| RULE_EXEC_PRIORITY | DELAY_RULE_PRIORITY |
| RULE_EXEC_REI_FILE_PATH | DELAY_RULE_REI_FILE_PATH |
| RULE_EXEC_STATUS | DELAY_RULE_STATUS |
| RULE_EXEC_TIME | DELAY_RULE_EXE_TIME |
| RULE_EXEC_USER_NAME | DELAY_RULE_USER_NAME |

As mentioned earlier, GenQuery2 exposes a subset of the GenQuery1 attributes. That means, many of the attributes were dropped from GenQuery2. Notable attributes include the following:

- `COLL_OWNER_NAME`
- `COLL_OWNER_ZONE`
- `DATA_OWNER_NAME`
- `DATA_OWNER_ZONE`

Logic relying on these attributes will need to be updated to use the permission attributes.

To get the full listing of attributes, use `iquery -c`. The listing can also be retrieved programmatically via the `rc_genquery2` API endpoint.

!!! Note

>     The results always reflect the attributes supported by the GenQuey2 implementation of the iRODS Catalog Provider.

Reviewer 2: ```suggestion
    The results always reflect the attributes supported by the GenQuery2 implementation of the iRODS Catalog Provider.
```

I'm a little confused by this sentence, specifically the "of the iRODS Catalog Provider" clause. Can you explain what that means?

Author: The GenQuery2 API goes to the catalog provider to build the list of supported attributes/column-names rather than the directly connected [consumer] server.

Author: Adjusting wording.

Author: Done.

# SQL functions in the ORDER-BY clause. 
iquery "select concat(COLL_NAME, concat('--', COLL_ACCESS_USER_NAME)) where COLL_NAME = '/tempZone/home/rods' order by concat(COLL_NAME, concat('--', COLL_ACCESS_USER_NAME))"

# Case-insensitive search.
iquery "select DATA_NAME where lower(DATA_NAME) = 'foo'"
iquery "select DATA_NAME where upper(DATA_NAME) = 'FOO'"

# Sub-string and integers.
iquery "select substr(DATA_NAME, 1, 3)"
```

!!! Note
    The behavior of aggregation operators and SQL functions depends on the database. GenQuery2 only acts as a passthrough.

### Metadata Queries

GenQuery2 allows users to write more targeted queries. This is especially helpful when dealing with metadata. Depending on the use-case, there may be times where you need to filter a resultset based on metadata attached to multiple iRODS entities (i.e. data objects, collections, resources, users). GenQuery2 makes this possible through the use of **SQL LEFT JOIN**. This happens automatically.

Here's an example demonstrating mixed metadata queries. _This is NOT possible with GenQuery1._

```sh
$ itouch foo
$ imeta add -d foo id 1000
$ itouch -R otherResc bar
$ imeta add -R otherResc speed fast

> $ iquery "select DATA_NAME where META_DATA_ATTR_VALUE = '1000' or META_RESC_ATTR_VALUE = 'fast'" | jq

Reviewer 2: A comment explaining what this query does may be helpful. Is it selecting only replicas which reside on resources with that metadata?

Author: It returns the data names which satisfy either of the following:
- There exists a data object with an attribute value of 1000 attached to it
- Or, a replica of a data object exists on a resource which has an attribute value of 'fast' attached to the resource

Both data objects are returned because both of them satisfy exactly one of the conditions.

Will add some words about that.

Author: Done.

Notice the use of `desc` and `asc`. Just like standard SQL, users can specify the sorting direction.

- Ascending order is specified by passing `asc` and is the _implicit_ default
- Descending order is specified by passing `desc`

Any number of attributes can be passed to the ORDER-BY clause as long as those attributes are defined in the PROJECTION clause.

### Escaping Bytes and Special Characters

Sometimes you may need to escape a specific byte (or character). GenQuery2 supports this through the use of hexadecimal encoding.

!!! Note
    This only applies to **string literals**. That is, the arguments wrapped in single quotes. Only **one byte** can be encoded at a time.

This is achieved by using `\xNN`, where _NN_ is the hexadecimal value of the byte. GenQuery2 decodes the byte before it reaches the database.

The following example demonstrates how to escape an exclamation mark.

```sh
iquery "select COLL_NAME, DATA_NAME where DATA_NAME = 'data\x21.txt'"
```

If the catalog contains a data object by the name, `data!.txt`, then it will be returned by GenQuery2.

Embedded single quotes can be escaped by adding another single quote, just like standard SQL. For example:

```sh

> iquery "select COLL_NAME, DATA_NAME where DATA_NAME = 'that''s my data.txt'

Reviewer 2: ```suggestion
iquery "select COLL_NAME, DATA_NAME where DATA_NAME = 'that''s my data.txt'"
```

Author: Done.

These features are great for relatively small datasets. However, as your datasets grow in size, you may want to build more sophisticated pagination solution.

### Resultsets and Pagination

Unlike GenQuery1, GenQuery2 does not provide built-in support for pagination. The resultset of a query will always be returned in its entirety. That means users fetching large amounts of data need to be careful of exhausting memory resources on the iRODS servers. Thankfully, GenQuery2 helps with this by defaulting to 256 rows if `LIMIT` is not defined in the query.

!!! Note
    `LIMIT` is provided as a convenience. It can be used in place of `FETCH FIRST N ROWS ONLY`, which is what's defined as part of the SQL standard.

While GenQuery2 doesn't provide built-in support for pagination, it does provide enough features for users to implement pagination efficiently. This is left as an exercise for the user.

### Other Notable Features

GenQuery2 supports the following additional features:

- Logical AND, OR, and NOT
- Grouping via parentheses
- SQL CAST
- Full range of relational operators: =, !=, <, <=, >, >=, LIKE, BETWEEN, IS [NOT] NULL
- SQL keywords are case-insensitive
- Federation

> GenQuery2 shares the same syntax as standard SQL.

Author: Consider moving this to the top of the GenQuery2 section.

Also, consider making examples for parents, etc.

Author: Done.

```
SELECT COLL_NAME, DATA_NAME WHERE DATA_RESC_NAME = 'replResc' and DATA_REPL_STATUS != '1'
```

## GenQuery2

> GenQuery2 is an experimental parser designed to give users more control over their queries.

Reviewer: we'll have to remove 'experimental' one day when GenQuery2 becomes default... so... perhaps 'new' is good enough here?

```suggestion
GenQuery2 is an new parser designed to give users more control over their queries.
```

Author: "new" is fine.

The note section immediately following that line explains the experimental-ness of the parser, so we don't lose anything by making the suggested edit.

Reviewer: right - i saw them later - i think there are two other mentions that we'd need to remove once we get to 'default GenQuery2'

Author: Correct.

Updated.

- Aggregation operators can be nested
- Aggregation operators can accept any number of input arguments
- SQL functions can be used in the PROJECTION, WHERE, and ORDER-BY clauses

Below are a few examples demonstrating each feature.

```sh
# SQL functions in the PROJECTION clause.
iquery "select concat(COLL_NAME, concat('--', COLL_ACCESS_USER_NAME)) where COLL_NAME = '/tempZone/home/rods'"

# SQL functions in the WHERE clause.
iquery "select COLL_NAME where concat(COLL_NAME, concat('--', COLL_ACCESS_USER_NAME)) = '/tempZone/home/rods--rods'"

# SQL functions in the ORDER-BY clause. 
iquery "select concat(COLL_NAME, concat('--', COLL_ACCESS_USER_NAME)) where COLL_NAME = '/tempZone/home/rods' order by concat(COLL_NAME, concat('--', COLL_ACCESS_USER_NAME))"

# Case-insensitive search.
iquery "select DATA_NAME where lower(DATA_NAME) = 'foo'"
iquery "select DATA_NAME where upper(DATA_NAME) = 'FOO'"

# Substrings and integers.
iquery "select substr(DATA_NAME, 1, 3)"
iquery "select substr(DATA_NAME, '1', '3')" # Produces the same output as the previous line.
```

!!! Note

>     The behavior of aggregation operators and SQL functions depends on the database. GenQuery2 only acts as a passthrough.

Reviewer: ```suggestion
    The behavior of aggregation operators and SQL functions depends on the database technology used by the iRODS catalog. GenQuery2 only acts as a passthrough.
```

Author: Done

!!! Note

>     Aggregation operators and SQL functions accept both integers and string literals. GenQuery2 only validates structure, not correctness of input arguments.

Reviewer: ```suggestion
    Aggregation operators and SQL functions accept both integers and string literals. GenQuery2 only validates structure, not correctness, of input arguments.
```

Author: Done

Sometimes you may need to escape a specific byte (or character). GenQuery2 supports this through the use of hexadecimal encoding.

!!! Note
    This only applies to **string literals**. That is, the arguments wrapped in single quotes. Only **one byte** can be encoded at a time.

This is achieved by using `\xNN`, where _NN_ is the hexadecimal value of the byte. GenQuery2 decodes the byte before it reaches the database.

The following example demonstrates how to escape an exclamation mark.

```sh
iquery "select COLL_NAME, DATA_NAME where DATA_NAME = 'data\x21.txt'"
```

If the catalog contains a data object by the name, `data!.txt`, then it will be returned by GenQuery2.

Embedded single quotes can be escaped by adding another single quote, just like standard SQL. For example:

```sh
iquery "select COLL_NAME, DATA_NAME where DATA_NAME = 'that''s my data.txt'"
```

GenQuery2 will notice the use of double single quotes and collapse them to one single quote before sending to the database. In the case of the example, that means `that''s` will be passed to the database as `that's`.

### Showing Duplicate Entries

By default, GenQuery2 removes duplicate entries from the resultset.

> If you need to view duplicate entries, add `no distinct` after the `select` keyword. For example:

Reviewer: is this standard SQL?  is there a better syntax for this?

Author: `no distinct` is not support by the SQL standard.

I pulled that forward during the early development stages of the parser and never went back to change it. Finding a better way to express that is on my TODO list. One way we could address is by doing something like what's described here:  https://github.com/irods/irods_api_plugin_genquery2/issues/42#issuecomment-1712525639. Alternatively, we introduce more support for the distinct keyword so that it behaves as defined by the SQL standard. I believe SQL allows users to apply the distinct keyword to one or more columns. Or maybe that's database specific. I'll verify and open an issue for it.

Reviewer: right, sticking to standard SQL is definitely the answer if plausible.  thanks.     please resolve this once you open a new issue.

Author: Created irods/irods#7944.

Reviewer: sold

GenQuery2 provides support for the GROUP-BY clause. The important thing to remember about GROUP-BY is that if the PROJECTION clause contains aggregation operators or SQL functions, the GROUP-BY clause may also require those same aggregation operators and/or SQL functions.

Here's an example that calculates the number of data objects in each collection.

```sh
iquery "select COLL_NAME, count(DATA_NAME) group by COLL_NAME"
```

### Offsets and Limiting the size of a resultset

By default, GenQuery2 will return a max of 256 rows if no limit is applied to the query.

You can change the size of the resultset by specifying a `LIMIT` or `FETCH FIRST N ROWS ONLY` clause. Both achieve the same outcome.

For example:

```sh
iquery "select COLL_NAME, DATA_NAME limit 1000"
```

!!! Note
    Keep in mind that GenQuery2 does NOT provide built-in pagination. It will always return the resultset in its entirety!

You can also apply an offset by using the `OFFSET` keyword.

```sh
iquery "select COLL_NAME, DATA_NAME order by COLL_NAME, DATA_NAME offset 2000 limit 1000"
```

Notice we're now also sorting the results. Without the ORDER-BY clause, the offset and limit add little value.

> These features are great for relatively small datasets. However, as your datasets grow in size, you may want to build more sophisticated pagination solution.

Reviewer: ```suggestion
These features are great for relatively small datasets. However, as your datasets grow in size, you may want to build a more sophisticated pagination solution.
```

Author: Done

!!! Note
    `OFFSET` and `LIMIT` only accept integers. Notice the examples do not wrap the integers in single quotes. Doing so will result in an error.

### Resultsets and Pagination

> Unlike GenQuery1, GenQuery2 does not provide built-in support for pagination. The resultset of a query will always be returned in its entirety. That means users fetching large amounts of data need to be careful of exhausting memory resources on the iRODS servers. Thankfully, GenQuery2 helps with this by defaulting to 256 rows if `LIMIT` is not defined in the query.

Reviewer: ```suggestion
Unlike GenQuery1, GenQuery2 does not provide built-in support for pagination. The resultset of a query will always be returned in its entirety. That means users fetching large amounts of data need to be careful of exhausting memory resources on the iRODS servers. GenQuery2 helps with this by defaulting to 256 rows if `LIMIT` is not defined in the query.
```

!!! Note
    `LIMIT` is provided as a convenience. It can be used in place of `FETCH FIRST N ROWS ONLY`, which is what's defined as part of the SQL standard.

While GenQuery2 doesn't provide built-in support for pagination, it does provide enough features for users to implement pagination efficiently. This is left as an exercise for the user.

### Limitations

> GenQuery2 is an experimental parser and does cover all use-cases supported by GenQuery1, yet.

Reviewer: ```suggestion
GenQuery2 is an experimental parser and does not yet cover all use-cases supported by GenQuery1.
```

Author: Done
